src/machine/machine_stm32_moder_gpio.go

```diff
@@ -0,0 +1,142 @@
+	GPIO_PULL_DOWN = 2
+)
+
+// Configure this pin with the given configuration.
+func (p Pin) configure(config PinConfig) {
+	// Configure the GPIO pin.
+	p.enableClock()
+	port := p.getPort()
+	pin := uint8(p) % 16
+	pos := pin * 2
+
+	switch config.Mode {
+
+	// GPIO
+	case PinInputFloating:
+		port.MODER.SetMasked(GPIO_MODE_INPUT, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_FLOATING, 0x3, pos)
+	case PinInputPulldown:
+		port.MODER.SetMasked(GPIO_MODE_INPUT, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_PULL_DOWN, 0x3, pos)
+	case PinInputPullup:
+		port.MODER.SetMasked(GPIO_MODE_INPUT, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_PULL_UP, 0x3, pos)
+	case PinOutput:
+		port.MODER.SetMasked(GPIO_MODE_GENERAL_OUTPUT, 0x3, pos)
+		port.OSPEEDR.SetMasked(GPIO_SPEED_HI, 0x3, pos)
+
+	// UART
+	case PinModeUartTX:
+		port.MODER.SetMasked(GPIO_MODE_ALTERNATIVE, 0x3, pos)
+		port.OSPEEDR.SetMasked(GPIO_SPEED_HI, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_PULL_UP, 0x3, pos)
+		// TODO: determine right alt-func for the peripheral in use
+		p.SetAltFunc(AF7_USART1_2_3)
+	case PinModeUartRX:
+		port.MODER.SetMasked(GPIO_MODE_ALTERNATIVE, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_FLOATING, 0x3, pos)
+		// TODO: determine right alt-func for the peripheral in use
+		p.SetAltFunc(AF7_USART1_2_3)
+
+	// I2C
+	case PinModeI2CScl:
+		port.MODER.SetMasked(GPIO_MODE_ALTERNATIVE, 0x3, pos)
+		port.OTYPER.SetMasked(GPIO_OUTPUT_MODE_OPEN_DRAIN, 0x1, pos)
+		port.OSPEEDR.SetMasked(GPIO_SPEED_LOW, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_FLOATING, 0x3, pos)
+		// TODO: determine right alt-func for the peripheral in use
+		p.SetAltFunc(AF4_I2C1_2_3)
+	case PinModeI2CSda:
+		port.MODER.SetMasked(GPIO_MODE_ALTERNATIVE, 0x3, pos)
+		port.OTYPER.SetMasked(GPIO_OUTPUT_MODE_OPEN_DRAIN, 0x1, pos)
+		port.OSPEEDR.SetMasked(GPIO_SPEED_LOW, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_FLOATING, 0x3, pos)
+		// TODO: determine right alt-func for the peripheral in use
+		p.SetAltFunc(AF4_I2C1_2_3)
+
+	// SPI
+	case PinModeSpiClk:
+		port.MODER.SetMasked(GPIO_MODE_ALTERNATIVE, 0x3, pos)
+		port.OSPEEDR.SetMasked(GPIO_SPEED_LOW, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_FLOATING, 0x3, pos)
+		// TODO: determine right alt-func for the peripheral in use
+		p.SetAltFunc(AF5_SPI1_SPI2)
+	case PinModeSpiMosi:
+		port.MODER.SetMasked(GPIO_MODE_ALTERNATIVE, 0x3, pos)
+		port.OSPEEDR.SetMasked(GPIO_SPEED_LOW, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_FLOATING, 0x3, pos)
+		// TODO: determine right alt-func for the peripheral in use
+		p.SetAltFunc(AF5_SPI1_SPI2)
+	case PinModeSpiMiso:
+		port.MODER.SetMasked(GPIO_MODE_ALTERNATIVE, 0x3, pos)
+		port.OSPEEDR.SetMasked(GPIO_SPEED_LOW, 0x3, pos)
+		port.PUPDR.SetMasked(GPIO_FLOATING, 0x3, pos)
+		// TODO: determine right alt-func for the peripheral in use
+		p.SetAltFunc(AF5_SPI1_SPI2)
+	}
+}
+
+// SetAltFunc maps the given alternative function to the I/O pin
+func (p Pin) SetAltFunc(af AltFunc) {
+	port := p.getPort()
+	pin := uint8(p) % 16
+	pos := pin * 4
+	if pin >= 8 {
```

Comment:
pos -= 8

Reply:
can you elaborate a little? I'm not quite sure where the `pos -= 8` should be going...

Reply:
Sorry, it should be `pin -= 8`.

```
	port := p.getPort()
	pin := uint8(p) % 16
	if pin >= 8 {
		pos := (pin - 8) * 4
		port.AFRH.SetMasked(...)
	} else {
		pos := pin * 4
		port.AFRL.SetMasked(...)
	}
```
When setting `AFRH`, the pin numbering starts with 8.

Reply:
The stm32f407 implementation uses only PA2 and PA3, so the bug has not been discovered. I was trying stm32f411 and found this.

Since there are many definitions (for example the USART interrupts) missing in the stm32f411 SVD files, I need some hacks to get it working. Probably I should wait until this patch is submitted.

In the meanwhile, I have discovered https://github.com/stm32-rs/stm32-rs, which fixed a lot of the bugs in the SVD files. Maybe we should generate device definitions from that repo.

Reply:
ah yes I see now. Thanks for the catch! I'll update the code. Good question on the SVD definitions; I was looking at getting a 401 and a 411 board so it's good to know there are gaps in the current implementation.

Reply:
@thinxer switching to stm32-rs for the SVD files may indeed be a good idea. It ensures we can re-use all the effort the Rust community put into fixing the SVD files :)